utils/trie_writer.hh

```diff
@@ -0,0 +1,1124 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+// Overview:
+//
+// This file implements a page-aware trie writer.
+//
+// The main class of this file, trie_writer, turns a sorted stream of keys into a stream of
+// trie nodes and padding which tries to balance the following goals:
+//
+// 1. Each node lies entirely within one page (guaranteed).
+// 2. Parents are located in the same page as their children (best-effort).
+// 3. Padding is minimized (best-effort).
+//
+// It does mostly what you'd expect a trie builder to do (whenever a new key is added,
+// branch out from the rightmost path and add a chain of new nodes, perhaps write some finished nodes to disk).
+// The complicated part is the size calculations needed to group parents and children together.
+//
+// For the most part we follow Cassandra's way of doing that.
+// See also:
+// https://github.com/apache/cassandra/blob/f9e2f1b219c0b570eb59d529da42722178608572/src/java/org/apache/cassandra/io/tries/IncrementalTrieWriterPageAware.java#L32
+// https://github.com/apache/cassandra/blob/f9e2f1b219c0b570eb59d529da42722178608572/src/java/org/apache/cassandra/io/sstable/format/bti/BtiFormat.md?plain=1#L347
+// The main difference is that our writer operates in units of character chains,
+// rather than on single characters.
+// With a node per character, the performance of a writer based on single characters
+// can get catastrophically bad for long keys.
+
+#include <seastar/util/log.hh>
+#include <map>
+#include <set>
+#include "utils/assert.hh"
+#include "utils/small_vector.hh"
+
+extern seastar::logger trie_logger;
+
+// Enables code which is useful for debugging during development,
+// but too expensive to be compiled into release builds (even if dynamically disabled).
+#define DEVELOPER_BUILD 0
```

Comment:
Please prefix it with TRIE_, the name is too globally scoped. If possible, make it a constexpr bool (just a preference on my part, not a requirement).

Reply:
Done.